content/post/101-rust-ipfs-unixfs.md

```diff
@@ -0,0 +1,152 @@
+tags: Rust, IPFS, UnixFS
+---
+
+UnixFS reading, or _exporting_, has landed in [Rust IPFS]! Funded by a
+Protocol Labs [devgrant], the main deliverables were the [`ipfs-unixfs`] crate,
+and the HTTP APIs `/get`, and `/cat`.
+
+[devgrant]: https://github.com/ipfs/devgrants/tree/master/open-grants/ipfs-rust/phase-2
+[Rust IPFS Append-only log #4]: https://medium.com/equilibriumco/rustipfs/home
+[Rust IPFS]: https://github.com/rs-ipfs/rust-ipfs
+[`ipfs-unixfs`]: https://crates.io/crates/ipfs-unixfs
+
+## The [ipfs-unixfs] crate
+
+You can review what UnixFS _is_ in our [previous post], but for now
+it's enough to know that a proper UnixFS implementation is basically
+about two things:
+
+1. Providing an interface to the data contained in the blocks. This is
+fairly straightforward.
+2. An ability to "walk" across multiple blocks via the MerkleDAG. This is
+much more difficult and, as such, most of the work went into finding
+a suitable abstraction.
+
+[previous post]: https://medium.com/equilibriumco/the-road-to-unixfs-f3cf5222b2ef
+[ipfs-unixfs]: https://crates.io/crates/ipfs-unixfs
+
+> To recap: [MerkleDAG is the outer protocol buffers description] for
+documents (or blocks) which can contain arbitrary inner bytes. Together
+with UnixFS blocks, these bytes correspond to [UnixFS messages].
+
+The operation needed when walking across multiple blocks is "loading the next
+block" or something along the lines of:
+
+```rust
+async fn get_block(..., cid: Cid) -> Result<impl AsRef<[u8]>, _>
+```
+
+However, we wanted to find a way to keep the IO of blocks separate from the
+implementation itself. This should have multiple benefits: any store integration - be it async or blocking - should be possible, and one should be able to
+compose the higher level operations out of the lower level pieces.
+
+So, instead of having an `ipfs_unixfs::GetBlock` or `ipfs_unixfs::BlockStore`
+trait, the `ipfs_unixfs::walk::Walker` currently has an API like this:
+
+```rust
+impl Walker {
+  fn new(start: Cid, root_name: &str) -> Walker;
+
+  fn pending_links(&self) -> (&Cid, impl Iterator<Item = &Cid> + '_) { ... }
+
+  fn continue_walk(self, next_block: &[u8]) -> Result<ContinuedWalk, _> { ... }
+}
+```
+
+The creation of `ipfs_unixfs::walk::Walker` is possible only with a root or
+start `Cid` and an optional path name to the root document. `Walker` tracks
+the path inside the graph for any entries.
+
+While `Walker::pending_links` may have a gnarly-looking signature, at
+the same time it communicates an important invariant of the API: whenever there
+exists a `Walker` value, there must be some pending links to load. This also means
+that the calling code does not have to deal with unwrapping an `Option<Cid>`
+from an `Iterator::next`, but can just get the next `Cid` as follows:
+
+```rust
+let (next, prefetchable_links) = walker.pending_links();
+```
+
+The iterator (`prefetchable_links`) in the tuple can be used to start
```

Comment:
who is this paragraph aimed at? what is my intended take away or action based on this info?

Reply:
Which paragraph, the one starting "While `Walker::pending_links` or the one starting "The iterator?"